src/concurrency.md

```diff
@@ -73,6 +74,63 @@ fn main() {
 }
 ```
 
+[ex-rayon-map-reduce]: #ex-rayon-map-reduce
+<a name="ex-rayon-map-reduce"></a>
+## Map-reduce in parallel
+
+[![rayon-badge]][rayon] [![cat-concurrency-badge]][cat-concurrency]
+
+This example uses [`rayon::filter`], [`rayon::map`], and [`rayon::reduce`]
+to calculate the coniditional average age of a vector of `Person` objects.
+
+[`rayon::filter`] allows us to (in parallel) conditionally include elements from
+a collection that satisfy the given predicate.  Similarly, [`rayon::map`] and
+[`rayon::reduce`] allow us to transform the filtered elements via a unary
+operation and reduce them to a single value via a given binary operation,
+respectively.
+
+```rust
+extern crate rayon;
+
+use rayon::prelude::*;
+
+struct Person {
+    age: u32,
+}
+
+fn main() {
+    let v: Vec<Person> = vec![
+        Person { age: 23 },
+        Person { age: 19 },
+        Person { age: 42 },
+        Person { age: 17 },
+        Person { age: 17 },
+        Person { age: 31 },
+        Person { age: 30 },
+    ];
+
+    let gt_30 = |x: u32| x > 30;
+    let num_over_30 = v.par_iter().filter(|&ref x| gt_30(x.age)).count();
+    let sum_over_30 = v.par_iter()
+        .filter(|&ref x| gt_30(x.age))
+        .map(|ref x| x.age)
+        .reduce(|| 0, |x, y| x + y);
+    
+    // Rayon also has a sum() method, which yields the same results as the
+    // reduction on the previous line
+    let alt_sum_30: u32 = v.par_iter()
+        .filter(|&ref x| gt_30(x.age))
+        .map(|ref x| x.age)
+        .sum();
+
+    let avg_over_30 = sum_over_30 as f32 / num_over_30 as f32;
```

Comment:
I'd probably move casts to above and store the variables as `f32` but it's a matter of taste.

Reply:
I run into a type inference issue with the `alt_sum_30` variable if I moved `as f32` to the line where `alt_sum_30` is initialized. So far the easiest way to get rid of that error is to make `alt_sum_30` a `u32` and then convert to `f32` later. Any suggestions?

Reply:
right `sum` is a template so type inference gets wonky here. 

```rust
 let alt_sum_30: f32 = v.par_iter()
        .filter(|x| gt_30(x.age))
        .map(|x| x.age as f32)
        .sum();
```
```rust
    let alt_sum_30 = v.par_iter()
        .filter(|x| gt_30(x.age))
        .map(|x| x.age as f32)
        .sum::<f32>();
```

or the cleanest solution, leaving is as it was, summed into u32 and casting later on ;)